src/librustc_middle/ty/codec.rs

```diff
@@ -60,6 +62,7 @@ where
     E: TyEncoder,
     M: for<'b> Fn(&'b mut E) -> &'b mut FxHashMap<T, usize>,
     T: EncodableWithShorthand,
+    usize: TryFrom<<T::Variant as DiscriminantKind>::Discriminant>,
```

Comment:
I don't think we can land the PR as-is with this code changed this way. Or at least I don't feel comfortable with it.

<hr/>

Let's think this through: we want to make sure the `discriminant` is less than `SHORTHAND_OFFSET`.

We can require discriminants are comparable (this might not be necessary because the trait might imply it):
```rust
<T::Variant as DiscriminantKind>::Discriminant: Ord
```

Then we just need to turn `SHORTHAND_OFFSET` into `<T::Variant as DiscriminantKind>::Discriminant`:

```rust
<T::Variant as DiscriminantKind>::Discriminant: Ord + TryFrom<usize>
```

Because `SHORTHAND_OFFSET` is constant, you can just do this and it should be equivalent AFAIK:
```rust
assert!(discriminant < SHORTHAND_OFFSET.try_into().unwrap())
```